### PR TITLE
GH-37643: [C++] Enhance arrow::Datum::ToString

### LINKED ISSUE #37643: [C++] Enhance arrow::Datum::ToString
### Describe the enhancement requested

As document says, `::arrow::Datum::ToString` only prints the "type" of tagged enum:

```
  /// \brief Return a string representation of the kind of datum stored.
  std::string ToString() const;
```

Can we enhance it and try call `ToString` for `RecordBatch`, `Table`, ...etc?

### Component(s)

C++

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Add print child fields for `arrow::Datum::ToString`. Because previously it just print a type.

### What changes are included in this PR?

Add detail in `arrow::Datum::ToString`

### Are these changes tested?

no

### Are there any user-facing changes?

Yes

* Closes: #37643

### CODE REVIEW

File: cpp/src/arrow/datum.cc [hunk 1]
```diff
@@ -182,15 +182,15 @@ std::string Datum::ToString() const {
     case Datum::NONE:
       return "nullptr";
     case Datum::SCALAR:
-      return "Scalar";
+      return "Scalar(" + scalar()->ToString() + ")";
     case Datum::ARRAY:
-      return "Array";
+      return "Array(" + make_array()->ToString() + ")";
```
Comment (mapleFU) [inline cpp/src/arrow/datum.cc:187]: `make_array` will generate a temp `::arrow::Array` object, don't know if we have a better way for this

**Comment by mapleFU** [comment]
```
/arrow/c_glib/test/test-table-datum.rb:52:in `test_to_string'
     49:   end
     50: 
     51:   def test_to_string
  => 52:     assert_equal("Table", @datum.to_s)
     53:   end
     54: 
     55:   def test_value
<"Table"> expected but was
<"Table(visible: bool\n" +
"----\n" +
"visible:\n" +
"  [\n" +
"    [\n" +
"      true,\n" +
"      false\n" +
"    ]\n" +
"  ]\n" +
")">
```

Seems this should also be fixed lol

**Comment by mapleFU** [comment]
@bkietz @kou Would you mind take a look?

**Review by bkietz — APPROVED** [review]
LGTM, thanks!

**Pull request merged**